### PR TITLE
Added the ability to read a test from a JSON file and the ability to answer them

### PR DESCRIPTION
**Pull request by vladpetrenciuc**

### CODE REVIEW

File: src/routes.js [hunk 1]
```diff
@@ -7,17 +7,29 @@ import {
...
 import Registration from "./registration";
 import TestDescription from "./testDescription";
+import TestStart from "./testStart";
+import Result from "./result";
 import UserContext from "./userContext";
 
 export default class RouteAndSwitch extends Component {
-  setUserState = ({ firstName, lastName, age }) => {
-    this.setState({ firstName: firstName, lastName: lastName, age: age });
+  setUserState = ({ firstName, lastName, age, testDifficulty, testTime, questions }) => {
```
Comment (dmfarcas) [inline src/routes.js:15]: Cred ca ai putea folosi direct context pentru a stoca datele necesare, nu? Pare ca le stochezi si in state, iar ulterior in context le trimiti doar cand faci acel submit.

File: src/testDescription.js [hunk 2]
```diff
@@ -8,17 +8,31 @@ export default class TestDescription extends Component {
...
-    this.handleSubmit = this.handleSubmit.bind(this); //this will probably be redone with arrow functions
   }
 
-  handleSubmit(event) {
+  handleChange = (event) => {
+    this.setState({ [event.target.name]: event.target.value });
+  }
+
+  handleSubmit = (context) => (event) => {
+    let testData = require("../testJSONs/difficulty" + this.state.testDifficulty + ".json");
```
Comment (dmfarcas) [inline src/testDescription.js]: const poate fi folosit aici

File: src/testDescription.js [hunk 3]
```diff
@@ -8,17 +8,31 @@ export default class TestDescription extends Component {
...
+    //this.setState({ testTime: testData.time.toString(), questions: testData.questions });
```
Comment (dmfarcas) [inline src/testDescription.js]: log

**Review by dmfarcas — APPROVED** [review]
As merge pe un functional refactor, pentru exercitiu. In loc de clase cu setState, ai putea folosi constante cu hooks. https://reactjs.org/docs/hooks-overview.html

File: src/result.js [hunk 4]
```diff
@@ -0,0 +1,23 @@
+import React, { Component } from "react";
+import UserContext from "./userContext";
+
+export default class Result extends Component{
```
Comment (dmfarcas) [inline src/result.js]: Aici, de exemplu, ai putea face const de pe acum, fara prea mult refactor :)

File: src/routes.js (hunk 1 above)
Comment (vladpetrenciuc) [reply src/routes.js:15]: Am construit un state aici, pt a putea folosi `<UserContext.Provider value={this.state}>`, pt ca daca nu am value imi da eroare la Consumer. Sau am inteles gresit la ce te referi?

**Review by adriangaro — APPROVED** [review]
Looks good to me, would recommend trying to look a bit into the functional component and hooks. Contexts are easier to use using hooks since you would have no need to have the consumer.

Also in class components context can be accessed without the consumer in render by adding the following in the class definition:
```js
class Result extends Component {
  public static contextType = UserContext;
  ...
```

Then you can access the context using `this.context` anywhere in the component. The only caveat is that you can consume only on context this way, but in your case is good enough.


Some useful links if ya wanna look:
[Hooks](https://reactjs.org/docs/hooks-intro.html)
[useContext](https://reactjs.org/docs/hooks-reference.html#useContext)

File: src/registration.js [hunk 5]
```diff
@@ -33,52 +40,38 @@ export default class Registration extends Component {
...
-            </form>
-          </div>)}
-
-        </UserContext.Consumer>
-      );
-    }
+  if (state.redirectToTest === true) {
+    return (
+      <Redirect to="/test" />
+    )
```
Comment (DanLepsa) [inline src/registration.js]: you should avoid rendering the Redirect. Instead use the useHistory hook from react router dom, then do history.push('/page')

File: src/registration.js [hunk 6]
```diff
@@ -1,22 +1,29 @@
...
+    setState(prevState => {
+      return { ...prevState, [event.target.name]: event.target.value }
+    });
+  }
 
-    this.handleChange = this.handleChange.bind(this);
-    this.handleSubmit = this.handleSubmit.bind(this); //this will probably be redone with arrow functions
+  const handleSubmit = (context) => (event) => {
+    context.setUserState({ firstName: state.firstName, lastName: state.lastName, age: state.age });
+    setState({ redirectToTest: true });
```
Comment (DanLepsa) [inline src/registration.js]: here you should do history.push. The redirectToTest state is not needed

File: src/registration.js [hunk 7]
```diff
@@ -33,52 +40,38 @@ export default class Registration extends Component {
...
-    }
+  if (state.redirectToTest === true) {
+    return (
+      <Redirect to="/test" />
+    )
+  } else {
+    return (
+      <div>
+        <h2>Registration</h2>
+        <form onSubmit={handleSubmit(context)}>
```
Comment (DanLepsa) [inline src/registration.js]: you don't need to send context to handle submit. when you get it from useContext() it's available for your component

File: src/testStart.js [hunk 8]
```diff
@@ -1,63 +1,56 @@
...
     } else {
-      this.setState({ question_number: this.state.question_number + 1 });
+      setState({ question_number: state.question_number + 1 });
     }
   }
 
-  buttonText = (context) => {
-    if (this.state.question_number < context.questions.length-1) {
+  const buttonText = (context) => {
+    if (this.state.question_number < context.questions.length - 1) {
```
Comment (DanLepsa) [inline src/testStart.js]: think you forgot to remove **this.** after converting to functional component

File: src/testStart.js [hunk 9]
```diff
@@ -1,63 +1,56 @@
...
+          {
+            Object.keys(context.questions[state.question_number]["answers"]).map((answer) => {
+              return (
+                <label key={answer}>
+                  <input type="radio" value={context.questions[state.question_number]["answers"][answer]} name="answer" />{context.questions[state.question_number]["answers"][answer]}
+                  <br />
+                </label>
+              )
+            })
+          }
```
Comment (DanLepsa) [inline src/testStart.js]: You should always try to split the render in smaller chunks. This part can be a function which returns the answers. Also using **object.item** notation is better than **object['item]**;

**Review by DanLepsa — APPROVED** [review]

**Pull request merged**